src/app/campaign/flight/flight-targets-form.component.ts

```diff
@@ -0,0 +1,144 @@
+      </div>
+    </fieldset>
+  `,
+  styleUrls: ['flight-targets-form.component.scss'],
+  providers: [
+    { provide: NG_VALUE_ACCESSOR, useExisting: FlightTargetsFormComponent, multi: true },
+    { provide: NG_VALIDATORS, useExisting: FlightTargetsFormComponent, multi: true }
+  ]
+})
+export class FlightTargetsFormComponent implements ControlValueAccessor {
+  private _targetOptions: InventoryTargets;
+  get targetOptions() {
+    return this._targetOptions;
+  }
+  @Input()
+  set targetOptions(targetOptions: InventoryTargets) {
+    this._targetOptions = targetOptions;
+    this.setCodeOptions(this.targets ? (this.targets.value as FlightTarget[]) : []);
+  }
+
+  targets: FormArray;
+  onChangeFn: (value: any) => void;
+  onTouchedFn: (value: any) => void;
+  typeOptions = [
+    { id: 'episode', label: 'Episode' },
+    { id: 'country', label: 'Country' }
+  ];
+  codeOptions: InventoryTarget[][];
+
+  newFlightTarget(target: FlightTarget) {
+    return new FormGroup({
+      type: new FormControl(target.type, Validators.required),
+      code: new FormControl(target.code, Validators.required),
+      exclude: new FormControl(target.exclude || false)
+    });
+  }
+
+  writeValue(targets: FlightTarget[]) {
+    this.setCodeOptions(targets);
+    this.targets = new FormArray((targets || []).map(this.newFlightTarget));
+    this.targets.valueChanges.subscribe(formTargets => {
+      this.setCodeOptions(formTargets);
+      this.onChangeFn(this.validTargetCodes(formTargets));
+    });
+  }
+
+  registerOnChange(fn: (value: any) => void) {
+    this.onChangeFn = fn;
+  }
+
+  registerOnTouched(fn: (value: any) => void) {
+    this.onTouchedFn = fn;
+  }
+
+  validate(_control: FormControl) {
+    return !this.targets || this.targets.valid ? null : { error: 'Invalid targets' };
+  }
+
+  addTarget() {
+    this.targets.markAsDirty();
+    this.targets.push(this.newFlightTarget({ type: '', code: '', exclude: false }));
+  }
+
+  removeTarget(index: number) {
+    this.targets.markAsDirty();
+    this.targets.removeAt(index);
+  }
+
+  private validTargetCodes(targets: FlightTarget[]) {
+    return targets.filter(({ type, code }, index) => {
+      if (!type || !code) {
+        return false;
+      } else if (!this.codeOptions[index].find(opt => opt.code === code)) {
+        // the type changed, and now the code is stale ... clear it!
+        this.targets
+          .at(index)
+          .get('code')
+          .setValue('');
+        return false;
+      } else {
+        return true;
+      }
+    });
+  }
+
+  private setCodeOptions(targets: FlightTarget[]) {
+    this.codeOptions = (targets || []).map(target => {
+      return this.filteredTargetOptions(target.type, target.code, targets);
+    });
+  }
+
+  private filteredTargetOptions(type: string, code: string, currentTargets: FlightTarget[]) {
+    return this.targetCodesForType(type).filter(opt => {
+      if (opt.type == type && opt.code == code) {
```

Comment:
what is this `==` business? ;) linter should catch that, but only if your editor is using it

Reply:
Hrmm.... yeah, what the heck?  I'll check that my linter is actually fixing these files.